juju/juju
# Hardcode ControllerBases and WorklodBases

This simplifies the signature and means we can remove a bunch of unused code

In Juju 4, now that we use bases everywhere, we no longer need a mechanism to sniff new series and/or bases in Juju - bases are deterministic

This means methods which return supported bases in different contexts can just return hard coded lists of bases we have tested Juju against (which can be overridden with the force flag for workloads)

## Checklist


## QA steps

Controller bases

```
$ juju bootstrap lxd lxd
Creating Juju controller "lxd" on lxd/localhost
Looking for packaged Juju agent version 4.0-beta3 for amd64
No packaged binary found, preparing local Juju agent binary
To configure your system to better support LXD containers, please see: https://documentation.ubuntu.com/lxd/en/latest/explanation/performance_tuning/
Launching controller instance(s) on localhost/localhost...
 - juju-928cb3-0 (arch=amd64)          
Installing Juju agent on bootstrap instance
Waiting for address
Attempting to connect to 10.219.211.131:22
Connected to 10.219.211.131
Running machine configuration script...
Bootstrap agent now started
Contacting Juju controller at 10.219.211.131 to verify accessibility...

Bootstrap complete, controller "lxd" is now available
Controller machines are in the "controller" model

Now you can run
	juju add-model <model-name>
to create a new model to deploy workloads.

$ juju status -m controller
Model       Controller  Cloud/Region         Version      Timestamp
controller  lxd         localhost/localhost  4.0-beta3.1  23:07:23+01:00

App         Version  Status  Scale  Charm            Channel     Rev  Exposed  Message
controller           active      1  juju-controller  4.0/stable   91  no       

Unit           Workload  Agent  Machine  Public address  Ports  Message
controller/0*  active    idle   0        10.219.211.131         

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.131  juju-928cb3-0  ubuntu@22.04      Running
```
```
$ juju bootstrap lxd focal --bootstrap-base ubuntu@20.04
Creating Juju controller "focal" on lxd/localhost
Looking for packaged Juju agent version 4.0-beta3 for amd64
No packaged binary found, preparing local Juju agent binary
To configure your system to better support LXD containers, please see: https://documentation.ubuntu.com/lxd/en/latest/explanation/performance_tuning/
Launching controller instance(s) on localhost/localhost...
 - juju-4e8ffe-0 (arch=amd64)          
Installing Juju agent on bootstrap instance
Waiting for address
Attempting to connect to 10.219.211.94:22
Connected to 10.219.211.94
Running machine configuration script...
Bootstrap agent now started
Contacting Juju controller at 10.219.211.94 to verify accessibility...

Bootstrap complete, controller "focal" is now available
Controller machines are in the "controller" model

Now you can run
	juju add-model <model-name>
to create a new model to deploy workloads.

$ juju status -m controller
Model       Controller  Cloud/Region         Version      Timestamp
controller  focal       localhost/localhost  4.0-beta3.1  23:07:48+01:00

App         Version  Status  Scale  Charm            Channel     Rev  Exposed  Message
controller           active      1  juju-controller  4.0/stable   91  no       

Unit           Workload  Agent  Machine  Public address  Ports  Message
controller/0*  active    idle   0        10.219.211.94          

Machine  State    Address        Inst id        Base          AZ  Message
0        started  10.219.211.94  juju-4e8ffe-0  ubuntu@20.04      Running
```
```
juju bootstrap lxd err --bootstrap-base ubuntu@21.04
Creating Juju controller "err" on lxd/localhost
ERROR failed to bootstrap model: use --force to override: ubuntu@21.04/stable not supported
```
```
$ juju bootstrap lxd err --bootstrap-base ubuntu@18.04
Creating Juju controller "err" on lxd/localhost
ERROR failed to bootstrap model: use --force to override: ubuntu@18.04/stable not supported
```

### Deploy some workloads

```
$ juju switch lxd
$ juju add-model m
$ juju deploy ubuntu
Deployed "ubuntu" from charm-hub charm "ubuntu", revision 24 in channel latest/stable on ubuntu@22.04/stable

$ juju deploy ubuntu focal --base ubuntu@20.04
Deployed "focal" from charm-hub charm "ubuntu", revision 24 in channel latest/stable on ubuntu@20.04/stable

(wait)
$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m      lxd         localhost/localhost  4.0-beta3.1  23:13:04+01:00

App     Version  Status  Scale  Charm   Channel        Rev  Exposed  Message
focal   20.04    active      1  ubuntu  latest/stable   24  no       
ubuntu  22.04    active      1  ubuntu  latest/stable   24  no       

Unit       Workload  Agent  Machine  Public address  Ports  Message
focal/0*   active    idle   1        10.219.211.201         
ubuntu/0*  active    idle   0        10.219.211.134         

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.134  juju-95cfe9-0  ubuntu@22.04      Running
1        started  10.219.211.201  juju-95cfe9-1  ubuntu@20.04      Running

$ juju deploy ubuntu --base ubuntu@18.04
ERROR base: ubuntu@18.04/stable not supported
ERROR failed to deploy charm "ubuntu"

$ juju deploy ubuntu --base ubuntu@21.04
ERROR selecting releases: charm or bundle not found for channel "", base "amd64/ubuntu/21.04"
available releases are:
...
```
